src/Operations.jl

```diff
@@ -643,25 +643,56 @@ function download_artifacts(ctx::Context, pkgs::Vector{PackageSpec}; platform::A
     return download_artifacts(ctx, pkg_roots; platform=platform, verbose=verbose)
 end
 
+"""
+    load_pkg_hook(hook_path::String)
+
+Given a path to a hook file (which should end in a function definition) include
+it into an anonymous module and return that function.
+"""
+function load_pkg_hook(hook_path::String)
+    m = Module(:__anon__)
+    f = Pkg.API.activate(dirname(dirname(hook_path))) do
+        Base.include(m, hook_path)::Function
+    end
```

Comment:
You need to performe the `LOAD_PATH` dance here.


What is the point of running with the package as the active project? Presumably you can't load the package anyway since we are still installing it? And this project might not even be valid (missing or inconsistent manifest, uninstalled dependencies etc) so you can't even expect to load other packages either.

Reply:
What is the `LOAD_PATH` dance?

The point is that perhaps someone's hook requires some fundamental piece of infrastructure; for instance, perhaps the routines that can tell us which CUDA artifact to download are in some `CUDAUtils` package that `CUDAnative` relies upon; it makes sense to me that the hook should be run with respect to that package's environment, even if it should be vanishingly rare for it to need that.

In what situations would there be invalid manifests or uninstalled dependencies?

Reply:
> What is the LOAD_PATH dance?

```
old_path = copy(LOAD_PATH)
empty!(LOAD_PATH)
push!(LOAD_PATH, "@")
try
   #...
finally
    empty!(LOAD_PATH)
    append!(LOAD_PATH, old_path)
end
```

> In what situations would there be invalid manifests or uninstalled dependencies?

Basically all the time? If the package doesn't commit the manifest it will error since it is missing, if it does come with a manifest it is unlikely to contain the same versions as Pkg decided upon and downloaded in the actual project.

> The point is that perhaps someone's hook requires some fundamental piece of infrastructure; for instance, perhaps the routines that can tell us which CUDA artifact to download are in some CUDAUtils package that CUDAnative relies upon; it makes sense to me that the hook should be run with respect to that package's environment, even if it should be vanishingly rare for it to need that.

Setting this up correctly is essentially what Pkg.build does. It creates a new project where you can load your own dependencies and starts a new Julia process with that env. 

I think that (i) the code you put in the hook file is completely self contained (not even using stdlibs) or (ii) we reuse Pkg.build, but the process prints out a program/config to `stdout` or something that can be read by the hook machinery in Artifacts.

Reply:
Yeah, I think you're right, let's just require it to be completely self-contained.  That's more in the spirit of a hook anyway.

Reply:
Another option would be to just mark all artifacts as lazy and just download the correct one in Pkg.build

Reply:
Heh, that is exactly what we're trying to avoid.  ;)  That's what is possible now, but it's a better experience for the ecosystem as a whole if we can know beforehand what will be downloaded.

Reply:
Right, but with this hook you can't "statically" decide that anyway since you need to run code, so it already is essentially `deps/build.jl` in disguise. Anyway, just a suggestion.

Reply:
You're right that it's not purely static, but once you have all package source available, it is a stateless operation to determine what artifacts need to be downloaded, which I think is as close to static as we can get without pushing more intelligence into base Julia itself.

Reply:
The `LOAD_PATH` dance doesn't really work because anything you have loaded in the current process will be used no matter what. That's why I suggested a separate process.